### ISSUE TITLE
fileset function should include optional ignore patterns

### ISSUE THREAD

**Issue by rvanider** [hashicorp/terraform#25074]
### Current Terraform Version

```
Terraform v0.12.25
```

### Use-cases

When using the `fileset()` function it is not possible to eliminate certain classes of files and folders from being processed.

The use cases that are desired:
- selectively ignore files and folders when using `for_each` and `fileset()`
- ability to apply separate resource definitions for `aws_s3_bucket_object` with different cache control and permissions while still referring to a single shared source of the files

### Attempted Solutions

The scenario that we are working with is the population of an S3 bucket with web site content where we want unique configuration for different subsets of files (eg. index.html) for control over the CloudFront caching behaviour.

Our solution, due to the absence of an ignore pattern, is to dissect the web site into multiple folders so that we can write multiple `aws_s3_bucket_object`resources,  each with varying configuration and differing by the folder it is sourcing from.

In order for this to work correctly and not have conflicting tracked resources (eg. same file uploaded to S3 by two different resources) is to make sure the folders have been properly sanitized. This is error prone and unique to each set of folders. It additionally, does not flow well when the source of the folder content is from a site build system such as webpack.

### Proposal

The proposed solution is to include an additional parameter to the `fileset()` function that identifies a set of ignore patterns that can be expressed in the same format as found in `.gitignore`, `.dockerignore`. Alternatives to this could be the addition of handling the detection and use of a `.tfignore` which provides the same functionality.

### References

n/a

**Comment by apparentlymart** [comment]
Hi @rvanider! Thanks for sharing this use-case.

In similar situations so far I've tended to address this by doing subsequent processing on the `fileset` result using language features like [`for` expressions](https://www.terraform.io/docs/configuration/expressions.html#for-expressions). For example, in [the directory template module](https://github.com/hashicorp/terraform-template-dir) we use [a sequence of transforms starting with a regular expression](https://github.com/hashicorp/terraform-template-dir/blob/a9c4ada84e6ecd892dfe7340743cfcdeb793d175/files.tf#L22-L30) to allow different treatment of files based on a suffix.

However, I know from that experience that a way to use the more concise "globbing" syntax familiar from shells and from `.gitignore` would be a much more convenient and readable way to get that done.

Since we try to keep functions orthogonal whereever possible, I'd consider meeting your use-case with a separate function that can do glob-style pattern matching on an existing file path string, which could be one returned from `fileset` or one obtained any other way. For example:

```hcl
locals {
  relevant_files = [ for fn in fileset("**") : fn if !globmatch(fn, ".*") ]
}
```

This `globmatch` function could potentially defined as variadic so that it can accept an arbitrary number of separate pattern strings, to keep things concise in more complex situations where otherwise there would be a chain of `globmatch` calls combined with `&&` or `||`.

Keeping this separate would be beneficial for situations like the aforementioned "template dir" module where the intent is to process _all_ of the files, but it needs to partition the set in to several smaller sets in order to do different processing on each of them. Doing `globmatch` as a separate step means we only need to scan the filesystem and produce the set once, and then we can filter it many times in-memory without rescanning the filesystem each time.

**Comment by rvanider** [comment]
I agree completely with the difficulty in trying to apply regular expressions to the results - it is a degree of complexity that is substantial relative to the task.

I think your proposed solution is both elegant and more than meets the use cases I could apply it to.

**Comment by luontola** [comment]
The string functions in the current Terraform version are already quite helpful for this:

```
resource "aws_s3_object" "html_public" {
  for_each = toset([
    for file in fileset(local.html_public_dir, "**") : file
    if !startswith(file, "js/cljs-runtime/") && !startswith(file, "fake-api/") && !endswith(file, ".DS_Store")
  ])
  # ...
}
```

**Comment by crw** [comment]
Thank you for your continued interest in this issue. 

[Terraform version 1.8](https://github.com/hashicorp/terraform/blob/v1.8.0-beta1/CHANGELOG.md) launches with support of provider-defined functions. It is now possible to implement your own functions! We would love to see this implemented as a provider-defined function. 

Please see the [provider-defined functions documentation](https://developer.hashicorp.com/terraform/plugin/framework/functions) to learn how to implement functions in your providers. If you are new to provider development, learn how to [create a new provider](https://developer.hashicorp.com/terraform/plugin/framework) with the Terraform Plugin Framework. If you have any questions, please visit the [Terraform Plugin Development category in our official forum](https://discuss.hashicorp.com/c/terraform-providers/tf-plugin-sdk/43). 

We hope this feature unblocks future function development and provides more flexibility for the Terraform community. Thank you for your continued support of Terraform!
